synnax/pkg/distribution/channel/channel_group.go

```diff
@@ -0,0 +1,273 @@
+			r.size++
+		} else {
+			c.insert(r)
+			r.start = localKeys[i]
+			r.size = 1
+		}
+	}
+	c.insert(r)
+}
+
+// RemoveKey removes key from c.
+func (c *ChannelGroup) RemoveKey(key Key) {
+	c.RemoveKeys(Keys{key})
+}
+
+// RemoveKeys removes all keys in keys from c.
+func (c *ChannelGroup) RemoveKeys(keys Keys) {
+	if len(keys) == 0 {
+		return
+	}
+
+	localKeys := make([]uint16, len(keys))
+	i := 0
+	for _, key := range keys {
+		if c.IsKeyIn(key) {
+			localKeys[i] = key.LocalKey()
+			i++
+		}
+	}
+	localKeys = localKeys[0:i]
+	if len(localKeys) == 0 {
+		return
+	}
+
+	r := channelRange{
+		start: localKeys[0],
+		size:  0,
+	}
+	for i, key := range localKeys {
+		if r.start+r.size == key {
+			r.size++
+		} else {
+			c.remove(r)
+			r.start = localKeys[i]
+			r.size = 1
+		}
+	}
+	c.remove(r)
+}
+
+// GetChannelsBeforeKey returns the number of channels in c before key. If key
+// is in c, that does not count towards the number of channels returned.
+func (c *ChannelGroup) GetChannelsBeforeKey(key Key) uint16 {
+	localKey := key.LocalKey()
+	counter := uint16(0)
+	for i := 0; i < len(*c) && (*c)[i].start < localKey; i++ {
+		if (*c)[i].end() >= localKey {
+			counter += localKey - (*c)[i].start
+		} else {
+			counter += (*c)[i].size
+		}
+	}
+	return counter
+}
+
+// IsKeyIn return true if the key is in c and false if key is not in c
+func (c ChannelGroup) IsKeyIn(key Key) bool {
+	localKey := key.LocalKey()
+	i := sort.Search(len(c), func(i int) bool { return c[i].start > localKey })
+	if i == 0 || c[i-1].end() < localKey {
+		return false
+	}
+	return true
+}
+
+// delete removes the element c[i] from c.
+func (c *ChannelGroup) delete(i int) {
+	(*c) = slices.Delete(*c, i, i+1)
+}
+
+// insert inserts r into c. insert also makes sure that c
+// is compressed to the minimum size possible.
+func (c *ChannelGroup) insert(r channelRange) {
+	if r.size == 0 {
+		return
+	}
+	if len(*c) == 0 {
+		*c = ChannelGroup{r}
+		return
+	}
+
+	// this is the first channelRange in c whose start is later than r's
+	i := sort.Search(len(*c), func(i int) bool { return (*c)[i].start > r.start })
+
+	// check if the channelRange to the left overlaps with r. If so, we remove the
+	// overlap from r and insert it again.
+	if i != 0 {
+		endOfBefore := (*c)[i-1].end()
+		if endOfBefore >= r.start {
+			overlap := endOfBefore - r.start + 1
+			if int(r.size)-int(overlap) <= 0 {
+				return
+			}
+			r.size -= overlap
+			r.start += overlap
+			c.insert(r)
+			return
+		}
+	}
+
+	// check if the channelRange to the right overlaps with r.
+	if i != len(*c) {
+		startOfNext := (*c)[i].start
+		if r.end() >= startOfNext {
+			if r.end() >= (*c)[i].end() {
+				(*c).delete(i)
+				c.insert(r)
+				return
+			}
+			overlap := r.end() - startOfNext + 1
+			r.size -= overlap
+			c.insert(r)
+			return
+		}
+	}
+
+	// At this point, r does not overlap at all with any of the
+	// channelRanges in c.
+
+	// This checks if we need to compress c

```

Comment:
Unify comment syntax: either they are all capitalized and have punctuation, or they are not. I think the standard is that if the comment is more than a few words, it should be a full sentence with capitalization and punctuation. @emilbon99